src/helpers/cloneChildren.js

```diff
@@ -1,38 +1,113 @@
 /*eslint func-style:0*/
 import React from 'react';
+import identity from './identity';
+import compose from './compose';
+import warning from 'warning';
 
 /**
- * Clones a child subtree, when we encounter a component that passes our
- * predicate pass it down additional props.
+ * Rule for cloning something at leaf level like some text
+ */
+const leafRule = {
+    predicate: (child) => typeof child !== 'object' || child === null,
+    clone: identity
+}
+
+/**
+ * Allows default recursion into an element that has children.
  *
- * @param  {Function} predicate A predicate function which recives the child
- * @param  {Function} getProps  A function which recives the component and
- * returns an object which gets merged into the props of the component
- * @param  {Function} children The children to iterate over
- * @return {Object} The cloned children
+ * @param {array} rules on how to clone individual elements
+ * @returns {Object} rule for cloning recursively
  */
-export default function cloneChildren(predicate, getProps, children) {
-    if (typeof children !== 'object' || children === null) {
-        return children;
+function createRecursiveRule(rules) {
+    return {
+        predicate: () => true,
+        clone: (child, childNames) => {
+            return React.cloneElement(child, {},
+                cloneChildren(rules, child.props && child.props.children, childNames));
+        }
     }
+}
 
-    return React.Children.map(children, (child) => {
-        if (typeof child !== 'object' || child === null) {
-            return child;
+/**
+ * A common function for cloning Errors element that takes care of injecting
+ * required error data
+ *
+ * @param {Object} errors aaaaa
+ * @param {Object} fieldErrors bbbb
+ * @return {Object} rule for cloning Errors element
+ */
+export function createErrorsRule({ errors = [], fieldErrors = {} }) {
+    return {
+        predicate: child => child.type && child.type.displayName === 'Errors',
+        clone: child => {
+            return React.cloneElement(
+                child,
+                {
+                    errors: errors,
+                    fieldErrors: fieldErrors
+                },
+                child.props && child.props.children
+            );
         }
+    }
+}
+
+/*
+ * Get extra properties for something we are going to weave our formable magic into.
+ */
+function getFormableComponentProperties(errors, fieldErrors, onSubmit, onChange) {
+    return (child, childNames) => {
+        warning(!child.ref, `Attempting to attach ref "${child.ref}" to "${child.props.name}" will be bad for your health`);
+        warning(childNames.indexOf(child.props.name) === -1, `Duplicate name "${child.props.name}" found. Duplicate fields will be ignored`);
+        childNames.push(child.props.name);
```

Comment:
Makes total sense, this is something that I'll probably look to address as a follow up given that it's how it was originally.

:bread: